discovery/watcher.go

```diff
@@ -161,14 +162,20 @@ func NewWatcher(ctx context.Context, config Config, log hclog.Logger) (*Watcher,
 }
 
 func (w *Watcher) Subscribe() <-chan State {
+	w.subscriptionMutex.Lock()
+	defer w.subscriptionMutex.Unlock()
+
 	ch := make(chan State, 1)
-	w.subcribers = append(w.subcribers, ch)
+	w.subscribers = append(w.subscribers, ch)
 	return ch
 }
 
 func (w *Watcher) notifySubscribers() {
+	w.subscriptionMutex.RLock()
```

Comment:
I think this is fine but I wanted to call out that this lock is held during all of the channel sends. Because the select used for sending has a default case the send will be non-blocking so I'm pretty sure this combo is actually ok.

Reply:
Yeah, this still could cause a problem if you have slow readers + lots of changes + additional subscribe calls though since we only have a buffered channel of size 1 and aren't doing anything like a `select`-based write...

Would it be worth just creating a copy of the array  prior to the iteration over it so we only hold the mutex lock while duping the subscriptions?

Reply:
nevermind, just re-read your comment and the code, you're right, it's already doing a select-based send.

Reply:
I don't think that's necessary. Creating a buffered channel of size one eliminates the need for a strict interlock sync between the producer and consumer goroutines and turns it more into a mailbox of sorts. Using the channel send with default select basically then means "don't notify me twice if I never even bothered to look at the first notification".

Now that said, it does mean that if you have a high churn of notifications for state changes and a consumer is laggy it'll be reading the FIRST notification not the most RECENT.

It could be updated this way:
```
func notify(ch chan State, state State) {
	// The loop here is just to attempt to re-send. There is a degenerate
	// case where you could starve the send if there are enough concurrent
	// notifications happening at once all trying to vie for the latest write.

	for {
		// Try to send if the channel is empty.
		select {
		case ch <- state:
			return
		default:
		}

		// Channel is not empty, drain prior state.
		select {
		case <-ch:
		default:
		}
	}
}
```